File: workspace_browser/shell_backends.js
'use strict';

/**
 * Platform-aware shell backends for the AILinux Loom desktop helper.
 *
 * Node port of local_workspace_client/shell_backends.py. Same contract, same
 * gate names: a backend must be available on this machine AND released by the
 * user before the AI is given a terminal. Unsandboxed backends are never
 * released implicitly.
 */

const { execFile } = require('node:child_process');
const fs = require('node:fs');
const path = require('node:path');

const SANDBOXED_BACKENDS = new Set(['bubblewrap', 'docker']);
const ENV_BACKEND = 'AILINUX_SHELL_BACKEND';
const ENV_DOCKER_IMAGE = 'AILINUX_SHELL_DOCKER_IMAGE';
const ENV_DOCKER_NETWORK = 'AILINUX_SHELL_DOCKER_NETWORK';

const BACKEND_LABELS = {
  bubblewrap: 'Linux console (bubblewrap sandbox)',
  linux: 'Linux console (native)',
  powershell: 'Windows PowerShell',
  cmd: 'Windows cmd.exe',
  macos: 'macOS console',
  docker: 'Docker environment',
};

function envValue(name) {
  return String(process.env[name] || '').trim();
}

function which(...names) {
  const exts = process.platform === 'win32' ? ['.exe', '.cmd', ''] : [''];
  for (const name of names) {
    if (!name) continue;
    if (name.includes(path.sep)) {
      try { fs.accessSync(name, fs.constants.X_OK); return name; } catch { continue; }
    }
    for (const dir of String(process.env.PATH || '').split(path.delimiter)) {
      if (!dir) continue;
      for (const ext of exts) {
        const candidate = path.join(dir, name + ext);
        try { fs.accessSync(candidate, fs.constants.X_OK); return candidate; } catch { /* next */ }
      }
    }
  }
  return '';
}

function backendAvailable(backend) {
  switch (backend) {
    case 'docker':
      if (!envValue(ENV_DOCKER_IMAGE)) return [false, `docker backend requires ${ENV_DOCKER_IMAGE}`];
      if (!which('docker')) return [false, 'docker executable not found'];
      return [true, `docker image ${envValue(ENV_DOCKER_IMAGE)}`];
    case 'bubblewrap':
      return which('bwrap') ? [true, 'bubblewrap sandbox'] : [false, 'bubblewrap (bwrap) not installed'];
    case 'powershell': {
      const found = which('pwsh', 'powershell');
      return found ? [true, `powershell at ${found}`] : [false, 'powershell not found'];
    }
    case 'cmd': {
      const found = envValue('COMSPEC') || which('cmd');
      return found ? [true, `cmd at ${found}`] : [false, 'cmd.exe not found'];
    }
    case 'macos': {
      const found = which('zsh', 'bash', 'sh');
      return found ? [true, `macOS shell at ${found}`] : [false, 'no zsh/bash/sh found'];
    }
    case 'linux': {
      const found = which('bash', 'sh');
      return found ? [true, `native shell at ${found}`] : [false, 'no bash/sh found'];
    }
    default:
      return [false, `unknown backend: ${backend}`];
  }
}

function detectBackend() {
  const pinned = envValue(ENV_BACKEND).toLowerCase();
  if (pinned) {
    if (!BACKEND_LABELS[pinned]) return ['', `unknown backend pinned via ${ENV_BACKEND}: ${pinned}`];
    const [ok, reason] = backendAvailable(pinned);
    return ok ? [pinned, reason] : ['', reason];
  }
  const order = [];
  if (envValue(ENV_DOCKER_IMAGE)) order.push('docker');
  if (process.platform === 'win32') order.push('powershell', 'cmd');
  else if (process.platform === 'darwin') order.push('macos');
  else order.push('bubblewrap', 'linux');

  let last = 'no shell backend available on this platform';
  for (const candidate of order) {
    const [ok, reason] = backendAvailable(candidate);
    if (ok) return [candidate, reason];
    last = reason;
  }
  return ['', last];
}

/** Release state lives in the main process; the renderer can never set it directly. */
const release = { granted: false, root: '', at: 0 };

function grantRelease(root) {
  release.granted = true;
  release.root = root ? path.resolve(root) : '';
  release.at = Date.now();
  return status();
}

function revokeRelease() {
  release.granted = false;
  release.root = '';
  release.at = 0;
  return status();
}

function status() {
  const [backend, detail] = detectBackend();
  const sandboxed = SANDBOXED_BACKENDS.has(backend);
  return {
    host: 'electron',
    platform: process.platform,
    backend,
    label: BACKEND_LABELS[backend] || '',
    available: Boolean(backend),
    sandboxed,
    released: Boolean(backend) && release.granted && Boolean(release.root),
    workspace: release.root,
    releasedAt: release.at,
    detail,
    releaseDetail: release.granted
      ? (release.root ? 'terminal released by the user' : 'released without a workspace folder')
      : 'terminal not released by the user',
  };
}

function posixArgs(shellPath, command) {
  const name = path.basename(shellPath);
  return (name === 'bash' || name === 'zsh')
    ? ['-o', 'pipefail', '-c', command]
    : ['-c', command];
}

function resolveWithin(root, candidate) {
  const base = path.resolve(root);
  const target = path.resolve(base, candidate || '.');
  const rel = path.relative(base, target);
  if (rel.startsWith('..') || path.isAbsolute(rel)) {
    throw new Error('path is outside the released workspace');
  }
  return { target, rel: rel.split(path.sep).filter(Boolean).join('/') };
}

function buildPlan(command, cwd) {
  const text = String(command || '').trim();
  if (!text) throw new Error('command is required');
  const state = status();
  if (!state.available) throw new Error(`no shell backend available: ${state.detail}`);
  if (!state.released) throw new Error(`shell not released: ${state.releaseDetail}`);

  const { target, rel } = resolveWithin(state.workspace, cwd);
  const workdir = '/workspace' + (rel ? `/${rel}` : '');

  if (state.backend === 'bubblewrap') {
    const argv = ['--die-with-parent', '--new-session', '--unshare-all',
      '--proc', '/proc', '--dev', '/dev', '--tmpfs', '/tmp',
      '--ro-bind', '/usr', '/usr', '--ro-bind', '/bin', '/bin', '--ro-bind', '/lib', '/lib'];
    if (fs.existsSync('/lib64')) argv.push('--ro-bind', '/lib64', '/lib64');
    argv.push('--bind', state.workspace, '/workspace', '--chdir', workdir,
      '--setenv', 'HOME', '/workspace', '--setenv', 'TMPDIR', '/tmp',
      '/bin/bash', '-o', 'pipefail', '-c', text);
    return { backend: 'bubblewrap', sandboxed: true, file: which('bwrap'), args: argv, cwd: undefined };
  }

  if (state.backend === 'docker') {
    const args = ['run', '--rm', '-i', '--network', envValue(ENV_DOCKER_NETWORK) || 'none',
      '-v', `${state.workspace}:/workspace`, '-w', workdir,
      envValue(ENV_DOCKER_IMAGE), '/bin/sh', '-c', text];
    return { backend: 'docker', sandboxed: true, file: 'docker', args, cwd: undefined };
  }

  if (state.backend === 'powershell') {
    const file = which('pwsh', 'powershell');
    return {
      backend: 'powershell', sandboxed: false, file,
      args: ['-NoLogo', '-NonInteractive', '-NoProfile', '-ExecutionPolicy', 'Bypass', '-Command', text],
      cwd: target,
    };
  }

  if (state.backend === 'cmd') {
    const file = envValue('COMSPEC') || which('cmd');
    return { backend: 'cmd', sandboxed: false, file, args: ['/d', '/c', text], cwd: target };
  }

  const file = state.backend === 'macos' ? which('zsh', 'bash', 'sh') : which('bash', 'sh');
  return { backend: state.backend, sandboxed: false, file, args: posixArgs(file, text), cwd: target };
}

function runShell({ command, cwd, timeout } = {}) {
  return new Promise((resolve) => {
    let plan;
    try {
      plan = buildPlan(command, cwd);
    } catch (error) {
      resolve({ ok: false, text: String(error.message || error), isError: true });
      return;
    }
    const limit = Math.max(1, Math.min(Number(timeout) || 120, 300)) * 1000;
    execFile(plan.file, plan.args, {
      cwd: plan.cwd, timeout: limit, maxBuffer: 4 * 1024 * 1024, windowsHide: true,
    }, (error, stdout, stderr) => {
      const parts = [];
      if (stdout) parts.push(`stdout:\n${stdout}`);
      if (stderr) parts.push(`stderr:\n${stderr}`);
      const code = error && typeof error.code === 'number' ? error.code : (error ? 1 : 0);
      if (error && error.killed) parts.push(`command timed out after ${limit / 1000}s`);
      parts.push(`exit_code=${code} backend=${plan.backend} sandboxed=${plan.sandboxed}`);
      const text = parts.join('\n').slice(0, 12000);
      resolve({ ok: code === 0, text, isError: code !== 0 });
    });
  });
}

module.exports = {
  SANDBOXED_BACKENDS, BACKEND_LABELS, ENV_BACKEND, ENV_DOCKER_IMAGE, ENV_DOCKER_NETWORK,
  detectBackend, backendAvailable, status, grantRelease, revokeRelease,
  buildPlan, runShell, resolveWithin, posixArgs, which,
};

File: local_workspace_client/shell_backends.py
"""Platform-aware shell backends for the TriForce / AILinux Loom workspace executor.

One canonical contract for every helper target. A backend turns a command into
an argv that runs against the shared workspace directory; Android via Termux,
Windows via PowerShell, Linux via bubblewrap or the native console, macOS via
zsh/bash and a Docker environment are backends -- not separate code paths and
not separate tool definitions.

Two independent gates decide whether the ``shell`` capability is offered at all:

1. availability -- does this machine actually provide the backend?
2. release     -- did the user explicitly hand the terminal to the AI?

Unsandboxed backends are never released implicitly. ``TRIFORCE_SHELL_RELEASE``
accepts ``auto`` (default: sandboxed backends only), ``off``, ``on`` (release
whatever is available) or an explicit backend id. ``TRIFORCE_SHELL_BACKEND``
pins one backend and disables auto-detection.
"""
from __future__ import annotations

import os
import platform
import shutil
import sys
from dataclasses import dataclass, field
from pathlib import Path

# Backends that confine the command to the workspace by construction. Only
# these may be released without an explicit opt-in from the user.
SANDBOXED_BACKENDS = frozenset({"bubblewrap", "docker"})

ENV_BACKEND = "TRIFORCE_SHELL_BACKEND"
ENV_RELEASE = "TRIFORCE_SHELL_RELEASE"
ENV_DOCKER_IMAGE = "TRIFORCE_SHELL_DOCKER_IMAGE"
ENV_DOCKER_NETWORK = "TRIFORCE_SHELL_DOCKER_NETWORK"

TERMUX_PREFIX = Path("/data/data/com.termux/files/usr")


class ShellUnavailable(RuntimeError):
    """No shell backend is available, or the terminal was not released."""


@dataclass(frozen=True)
class ShellPlan:
    backend: str
    argv: list[str]
    sandboxed: bool
    cwd: str | None = None
    env: dict[str, str] = field(default_factory=dict)


def _env(name: str, default: str = "") -> str:
    return str(os.environ.get(name) or default).strip()


def is_termux() -> bool:
    """Android/Termux, detected without assuming the app is the caller."""
    if _env("TERMUX_VERSION"):
        return True
    if "com.termux" in _env("PREFIX"):
        return True
    return (TERMUX_PREFIX / "bin").is_dir()


def is_android() -> bool:
    if is_termux():
        return True
    return "android" in platform.platform().lower() or Path("/system/build.prop").exists()


def _which(*names: str) -> str:
    for name in names:
        found = shutil.which(name)
        if found:
            return found
    return ""


def _termux_bin(name: str) -> str:
    candidate = TERMUX_PREFIX / "bin" / name
    return str(candidate) if candidate.exists() else ""


def _docker_image() -> str:
    return _env(ENV_DOCKER_IMAGE)


def detect_backend() -> tuple[str, str]:
    """Return ``(backend_id, reason)``; backend_id is empty when unavailable."""
    pinned = _env(ENV_BACKEND).lower()
    if pinned:
        if pinned not in BACKEND_LABELS:
            return "", f"unknown backend pinned via {ENV_BACKEND}: {pinned}"
        ok, reason = _backend_available(pinned)
        return (pinned, reason) if ok else ("", reason)

    order: list[str] = []
    if _docker_image():
        order.append("docker")
    if is_android():
        order.append("termux")
    elif sys.platform.startswith("win"):
        order += ["powershell", "cmd"]
    elif sys.platform == "darwin":
        order.append("macos")
    else:
        order += ["bubblewrap", "linux"]

    last = "no shell backend available on this platform"
    for candidate in order:
        ok, reason = _backend_available(candidate)
        if ok:
            return candidate, reason
        last = reason
    return "", last


def _backend_available(backend: str) -> tuple[bool, str]:
    if backend == "docker":
        if not _docker_image():
            return False, f"docker backend requires {ENV_DOCKER_IMAGE}"
        if not _which("docker"):
            return False, "docker executable not found"
        return True, f"docker image {_docker_image()}"
    if backend == "bubblewrap":
        if not _which("bwrap"):
            return False, "bubblewrap (bwrap) not installed"
        return True, "bubblewrap sandbox"
    if backend == "termux":
        if _termux_bin("bash") or _termux_bin("sh"):
            return True, "termux shell"
        if Path("/system/bin/sh").exists():
            return True, "android system shell (termux not installed)"
        return False, "no Termux prefix and no /system/bin/sh"
    if backend == "powershell":
        found = _which("pwsh", "powershell", "powershell.exe")
        return (True, f"powershell at {found}") if found else (False, "powershell not found")
    if backend == "cmd":
        found = _env("COMSPEC") or _which("cmd.exe")
        return (True, f"cmd at {found}") if found else (False, "cmd.exe not found")
    if backend == "macos":
        found = _which("zsh", "bash", "sh")
        return (True, f"macOS shell at {found}") if found else (False, "no zsh/bash/sh found")
    if backend == "linux":
        found = _which("bash", "sh")
        return (True, f"native shell at {found}") if found else (False, "no bash/sh found")
    return False, f"unknown backend: {backend}"


BACKEND_LABELS = {
    "bubblewrap": "Linux console (bubblewrap sandbox)",
    "linux": "Linux console (native)",
    "termux": "Android (Termux)",
    "powershell": "Windows PowerShell",
    "cmd": "Windows cmd.exe",
    "macos": "macOS console",
    "docker": "Docker environment",
}


def release_mode() -> str:
    value = _env(ENV_RELEASE, "auto").lower()
    return value if value in {"auto", "on", "off"} or value in BACKEND_LABELS else "auto"


def is_released(backend: str) -> tuple[bool, str]:
    """Has the user handed this backend's terminal to the AI?"""
    if not backend:
        return False, "no backend"
    mode = release_mode()
    if mode == "off":
        return False, f"terminal access disabled ({ENV_RELEASE}=off)"
    if mode == "on":
        return True, "released for all available backends"
    if mode in BACKEND_LABELS:
        if mode == backend:
            return True, f"released explicitly for {backend}"
        return False, f"{ENV_RELEASE} released '{mode}', not '{backend}'"
    # auto
    if backend in SANDBOXED_BACKENDS:
        return True, "sandboxed backend, released automatically"
    return False, (
        f"'{backend}' runs unsandboxed; set {ENV_RELEASE}=on or {ENV_RELEASE}={backend} "
        "to hand the terminal to the AI"
    )


def shell_backend_status(root: Path | None = None) -> dict[str, object]:
    """Machine-readable state for workspace_info, the helper UI and logs."""
    backend, reason = detect_backend()
    released, release_reason = (is_released(backend) if backend else (False, reason))
    return {
        "backend": backend,
        "label": BACKEND_LABELS.get(backend, ""),
        "available": bool(backend),
        "sandboxed": backend in SANDBOXED_BACKENDS,
        "released": bool(backend) and released,
        "release_mode": release_mode(),
        "detail": reason,
        "release_detail": release_reason,
        "platform": "android" if is_android() else sys.platform,
        "workspace": str(root) if root else "",
    }


def _posix_argv(shell: str, command: str) -> list[str]:
    # pipefail is a bash/zsh feature; plain sh (dash) would abort on it.
    name = Path(shell).name
    if name in {"bash", "zsh"}:
        return [shell, "-o", "pipefail", "-c", command]
    return [shell, "-c", command]


def build_shell_plan(command: str, *, root: Path, work: Path) -> ShellPlan:
    """Build the concrete invocation, or raise ``ShellUnavailable``."""
    command = str(command or "").strip()
    if not command:
        raise ShellUnavailable("command is required")

    backend, reason = detect_backend()
    if not backend:
        raise ShellUnavailable(f"no shell backend available: {reason}")
    released, release_reason = is_released(backend)
    if not released:
        raise ShellUnavailable(f"shell not released: {release_reason}")

    root = Path(root).resolve()
    work = Path(work).resolve()
    rel = work.relative_to(root)
    rel_posix = rel.as_posix() if rel.parts else ""
    sandboxed = backend in SANDBOXED_BACKENDS

    if backend == "bubblewrap":
        bwrap = _which("bwrap")
        argv = [bwrap, "--die-with-parent", "--new-session", "--unshare-all",
                "--proc", "/proc", "--dev", "/dev", "--tmpfs", "/tmp",
                "--ro-bind", "/usr", "/usr", "--ro-bind", "/bin", "/bin",
                "--ro-bind", "/lib", "/lib"]
        if Path("/lib64").exists():
            argv += ["--ro-bind", "/lib64", "/lib64"]
        argv += ["--bind", str(root), "/workspace",
                 "--chdir", "/workspace" + (("/" + rel_posix) if rel_posix else ""),
                 "--setenv", "HOME", "/workspace", "--setenv", "TMPDIR", "/tmp",
                 "/bin/bash", "-o", "pipefail", "-c", command]
        return ShellPlan(backend, argv, True, cwd=None, env=dict(os.environ))

    if backend == "docker":
        image = _docker_image()
        argv = ["docker", "run", "--rm", "-i",
                "--network", _env(ENV_DOCKER_NETWORK, "none"),
                "-v", f"{root}:/workspace",
                "-w", "/workspace" + (("/" + rel_posix) if rel_posix else ""),
                image, "/bin/sh", "-c", command]
        return ShellPlan(backend, argv, True, cwd=None, env=dict(os.environ))

    env = dict(os.environ)

    if backend == "termux":
        shell = _termux_bin("bash") or _termux_bin("sh") or "/system/bin/sh"
        if str(TERMUX_PREFIX) not in env.get("PATH", ""):
            env["PATH"] = f"{TERMUX_PREFIX / 'bin'}:{env.get('PATH', '')}".rstrip(":")
        env.setdefault("PREFIX", str(TERMUX_PREFIX))
        return ShellPlan(backend, _posix_argv(shell, command), False, cwd=str(work), env=env)

    if backend == "powershell":
        shell = _which("pwsh", "powershell", "powershell.exe")
        argv = [shell, "-NoLogo", "-NonInteractive", "-NoProfile",
                "-ExecutionPolicy", "Bypass", "-Command", command]
        return ShellPlan(backend, argv, False, cwd=str(work), env=env)

    if backend == "cmd":
        shell = _env("COMSPEC") or _which("cmd.exe")
        return ShellPlan(backend, [shell, "/d", "/c", command], False, cwd=str(work), env=env)

    shell = _which("zsh", "bash", "sh") if backend == "macos" else _which("bash", "sh")
    return ShellPlan(backend, _posix_argv(shell, command), False, cwd=str(work), env=env)
